File: src/backdrop.ts
import { invoke } from "@tauri-apps/api/core";
import { useEffect, useState } from "react";

import type { Theme } from "./theme";

/**
 * The colour behind the webview.
 *
 * A webview is a rectangle drawn on top of a native view, and that view has a
 * colour of its own that the page cannot declare. It shows wherever the page is
 * not painted yet — most visibly while a device is being turned, because the
 * window changes shape a moment before the page has been laid out again inside
 * it. Unset, it is white, and a white sheet flashing behind a dark wallet is
 * exactly what somebody rotating a tablet sees.
 *
 * **The stylesheet stays the only place a colour is written down.** This reads
 * back what the page is actually painting and hands that same value to the
 * window, rather than keeping a second copy of the palette over here that would
 * be wrong the first time anybody edited the first one.
 *
 * It is read from `body` rather than from the `--bg` token because a custom
 * property answers with whatever text was typed into it, while a computed
 * background answers with the colour the platform is really drawing — which is
 * the thing that has to match.
 *
 * On a computer the window's own chrome is sent the theme as well, because a
 * title bar the system draws light over a page that is dark is the same flash
 * made permanent — see `src-tauri/src/backdrop.rs`.
 *
 * It goes through this application's own command and not through Tauri's
 * `setBackgroundColor`, which is registered for desktop alone and rejects on a
 * phone — the one place any of this matters.
 */

/** Whichever way the device is leaning, when nobody has chosen for it. */
function useSystemScheme(): "dark" | "light" {
  const [scheme, setScheme] = useState<"dark" | "light">(() =>
    window.matchMedia("(prefers-color-scheme: dark)").matches
      ? "dark"
      : "light",
  );

  useEffect(() => {
    const media = window.matchMedia("(prefers-color-scheme: dark)");
    const answer = () => setScheme(media.matches ? "dark" : "light");

    media.addEventListener("change", answer);
    return () => media.removeEventListener("change", answer);
  }, []);

  return scheme;
}

/** What the page is painting behind everything, as three or four numbers. */
function painted(): [number, number, number, number] | null {
  const computed = window.getComputedStyle(document.body).backgroundColor;
  const parts = computed.match(/[\d.]+/g);
  if (parts === null || parts.length < 3) {
    return null;
  }

  const [red, green, blue, alpha] = parts.map(Number);

  // An alpha the page did not ask for would let the white through again, so a
  // missing one is opaque rather than nothing.
  return [
    red,
    green,
    blue,
    alpha === undefined ? 255 : Math.round(alpha * 255),
  ];
}

/**
 * Keep the window's colour the same as the page's.
 *
 * `theme` and the system's own answer are both watched, because either can move
 * the palette: somebody choosing dark, and somebody's phone deciding it is
 * evening, arrive here the same way.
 *
 * `scanning` pauses it. The scanner makes the webview transparent so the camera
 * behind it can be seen, and a colour pushed while it is open would paint over
 * the picture.
 */
export function useBackdrop(theme: Theme, scanning: boolean): void {
  const scheme = useSystemScheme();

  useEffect(() => {
    if (scanning) {
      return;
    }

    // Read after the theme has reached the document: `getComputedStyle` settles
    // the pending style change, so this is the palette that is about to be on
    // screen and not the one leaving it.
    const colour = painted();
    if (colour === null) {
      return;
    }

    const [red, green, blue, alpha] = colour;

    // `system` is the absence of a choice, and the window is handed back to the
    // system in the same terms rather than told which way the system leans:
    // the window already follows that on its own, and telling it would stop it.
    const chosen = theme === "system" ? null : theme;

    // A platform that will not take a colour keeps the one it has. The page is
    // painted the same either way — this is only what shows around it, and it
    // is not worth a message to somebody who cannot act on it.
    void invoke("backdrop_set", {
      red,
      green,
      blue,
      alpha,
      theme: chosen,
    }).catch(() => undefined);
  }, [theme, scheme, scanning]);
}

File: src-tauri/src/backdrop.rs
//! The colour behind the webview, and the appearance of the window around it.
//!
//! A webview is a rectangle drawn on top of a native view, and that view has a
//! colour of its own that the page cannot declare. It shows wherever the page is
//! not painted yet — most visibly while a tablet is being turned, because the
//! window changes shape a moment before the page has been laid out again inside
//! it. Left alone it is white, and a white sheet flashing behind a dark wallet
//! is what somebody rotating a device sees.
//!
//! On a computer the window itself is a second surface with the same problem,
//! and a more visible one: the title bar. Since macOS Tahoe a standard title bar
//! is a slab of the system's own grey, drawn in whichever of light or dark the
//! window is in, and over a dark page that is a light band with a hard edge
//! under it. The window is declared with `titleBarStyle: "Transparent"` in
//! `tauri.conf.json`, which makes the bar show the window's *background colour*
//! instead — the page stays below it, nothing is drawn underneath the title —
//! and so the window is painted the page's colour here too, not only the
//! webview. `tauri.conf.json` gives it the light background so the first frame
//! is not white; on macOS the webview setter is not implemented, so without this
//! the bar would keep that light colour under a dark page.
//!
//! The window is also told which appearance it is wearing, because the bar's
//! text and controls are drawn by the system in the appearance it believes the
//! window to be in — the system's setting, until the wallet says otherwise —
//! and a dark title over a dark page is somebody who chose dark against a light
//! system.
//!
//! **The value is not written down here.** The stylesheet is the only place this
//! project names a colour, so the interface reads back what it is actually
//! painting and hands that across — see `src/backdrop.ts`. This side only
//! forwards it, which is why the argument is four numbers and not a palette.
//!
//! It is a command of this application's rather than the ones Tauri already has,
//! because Tauri's are registered `#[cfg(desktop)]`: `webview.setBackgroundColor`
//! from the interface rejects on a phone, which is the one place it is needed.
//! The Rust API underneath has no such restriction and reaches both platforms.

use serde::{Serialize, Serializer};
use tauri::utils::config::Color;
use tauri::{Runtime, Theme, Webview};

/// What can go wrong, as a code rather than prose.
#[derive(Debug, Clone, Copy)]
pub enum BackdropError {
    /// The platform would not take the colour.
    Refused,
}

impl BackdropError {
    const fn code(self) -> &'static str {
        match self {
            Self::Refused => "backdrop_refused",
        }
    }
}

impl Serialize for BackdropError {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_str(self.code())
    }
}

/// Paint the window behind the page the same colour the page is painting, and
/// dress its chrome to match.
///
/// `theme` is the palette the page is on — `light` or `dark` — or nothing when
/// the wallet is following the system, in which case the window is handed back
/// to the system too. It is the interface's choice and not `prefers-color-scheme`
/// that is sent, because the two differ exactly when somebody picked a theme
/// against the system's, and that is the case the chrome has to be told about.
///
/// Answers the calling webview rather than looking one up by name: the wallet
/// has exactly one, and asking for it by label would be a second place that has
/// to agree with `tauri.conf.json`.
#[tauri::command]
pub fn backdrop_set<R: Runtime>(
    webview: Webview<R>,
    red: u8,
    green: u8,
    blue: u8,
    alpha: u8,
    theme: Option<Theme>,
) -> Result<(), BackdropError> {
    let colour = Some(Color(red, green, blue, alpha));

    // The window before the webview: on the desktops it is the one that shows,
    // and a webview that will not take the colour is not worth losing it over.
    #[cfg(desktop)]
    {
        let window = webview.window();
        window
            .set_background_color(colour)
            .map_err(|_| BackdropError::Refused)?;
        window
            .set_theme(theme)
            .map_err(|_| BackdropError::Refused)?;
    }
    #[cfg(not(desktop))]
    let _ = theme;

    webview
        .set_background_color(colour)
        .map_err(|_| BackdropError::Refused)
}
